### PR TITLE
CLOUDP-97429: Create one new live migration

### PR DESCRIPTION
**Pull request by fmenezes**
## Proposed changes

Adding the command `mongocli atlas liveMigrations create`

_Jira ticket:_ CLOUDP-97429

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have updated [e2e/E2E-TESTS.md](e2e/E2E-TESTS.md) (if a new command or e2e test has been added)
- [x] I have run `make fmt` and formatted my code

### CODE REVIEW

File: internal/store/live_migration_validations.go [hunk 1]
```diff
@@ -28,10 +28,6 @@ type LiveMigrationValidationsCreator interface {
 	CreateValidation(string, *atlas.LiveMigration) (*atlas.Validation, error)
 }
 
-type LiveMigrationValidationsStore interface {
```
Comment (fmenezes) [inline internal/store/live_migration_validations.go:31]: Not used

File: internal/cli/atlas/livemigrations/validation/create.go [hunk 2]
```diff
@@ -170,14 +170,14 @@ func CreateBuilder() *cobra.Command {
 	cmd.Flags().BoolVar(&opts.sourceSSL, flag.LiveMigrationSourceSSL, false, usage.LiveMigrationSourceSSL)
 	cmd.Flags().StringVar(&opts.sourceCACertificatePath, flag.LiveMigrationSourceCACertificatePath, "", usage.LiveMigrationSourceCACertificatePath)
 	cmd.Flags().BoolVar(&opts.sourceManagedAuthentication, flag.LiveMigrationSourceManagedAuthentication, false, usage.LiveMigrationSourceManagedAuthentication)
-	cmd.Flags().StringVar(&opts.destinationClusterName, flag.LiveMigrationDestinationClusterName, "", usage.LiveMigrationDestinationClusterName)
-	cmd.Flags().StringVar(&opts.ProjectID, flag.LiveMigrationDestinationProjectID, "", usage.LiveMigrationDestinationProjectID)
+	cmd.Flags().StringVar(&opts.destinationClusterName, flag.ClusterName, "", usage.LiveMigrationDestinationClusterName)
+	cmd.Flags().StringVar(&opts.ProjectID, flag.ProjectID, "", usage.LiveMigrationDestinationProjectID)
```
Comment (fmenezes) [inline internal/cli/atlas/livemigrations/validation/create.go:174]: The spec had an issue I'm fixing it

cc @colm-quinn

Comment (colm-quinn) [reply internal/cli/atlas/livemigrations/validation/create.go:174]: Should we not use the new `LiveMigrationDestinationClusterName` etc here, as the description is more accurate/helpful to users?

File: internal/cli/atlas/livemigrations/create.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	}
+	return nil
+}
+
+// mongocli atlas liveMigrations|lm create --clusterName clusterName --migrationHosts hosts --sourceClusterName clusterName --sourceProjectId projectId [--sourceSSL] [--sourceCACertificatePath path] [--sourceManagedAuthentication] [--sourceUsername userName] [--sourcePassword password] [--drop] [--projectId projectId].
+func CreateBuilder() *cobra.Command {
+	opts := &CreateOpts{}
+	cmd := &cobra.Command{
+		Use:   "create",
+		Short: "Create one new validation request.",
```
Comment (gssbzn) [inline internal/cli/atlas/livemigrations/create.go:150]: copy/paste?

File: internal/cli/atlas/livemigrations/create.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	destinationDropEnabled      bool
+	migrationHosts              []string
+	sourceCACertificatePath     string
+	sourceClusterName           string
+	sourceProjectID             string
+	sourceSSL                   bool
+	sourceManagedAuthentication bool
+	sourceUsername              string
+	sourcePassword              string
+	store                       store.LiveMigrationCreator
```
Comment (gssbzn) [inline internal/cli/atlas/livemigrations/create.go:45]: I left a note on the validation PR, when adding this command given the behave exactly the same with a different store to have a `CloudMigrationOpts` or something that both could reuse

File: internal/cli/atlas/livemigrations/create.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+	destinationClusterName      string
+	destinationDropEnabled      bool
+	migrationHosts              []string
+	sourceCACertificatePath     string
+	sourceClusterName           string
+	sourceProjectID             string
+	sourceSSL                   bool
+	sourceManagedAuthentication bool
+	sourceUsername              string
+	sourcePassword              string
```
Comment (colm-quinn) [inline internal/cli/atlas/livemigrations/create.go:44]: A lot of this is common with the validation data - we should break it out into a common struct and reuse now that it's been used in a second place.

File: internal/cli/atlas/livemigrations/validation/create.go (hunk 2 above)
Comment (fmenezes) [reply internal/cli/atlas/livemigrations/validation/create.go:174]: the descriptions are changed, we are using the ProjectID instead for configuration wise and `clusterName` flag is widely used already in the cli.

There was a gap in the spec we basically documented both (copy/paste issue). The help will display the word `destination` to correctly explain it.

File: internal/cli/atlas/livemigrations/create.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+		return fmt.Errorf("--%s and --%s are exclusive", flag.LiveMigrationSourceManagedAuthentication, flag.LiveMigrationSourceUsername)
+	}
+	if !opts.sourceManagedAuthentication && opts.sourceUsername == "" {
+		return fmt.Errorf("MongoDB Automation is not managing authentication, --%s must be set", flag.LiveMigrationSourceUsername)
+	}
+	if opts.sourceCACertificatePath != "" {
+		opts.sourceSSL = true
+	}
+	return nil
+}
```
Comment (colm-quinn) [inline internal/cli/atlas/livemigrations/create.go:143]: Same as the validation point - we should do this in one place and reuse if we can.

File: internal/cli/atlas/livemigrations/validation/create.go (hunk 2 above)
Comment (colm-quinn) [reply internal/cli/atlas/livemigrations/validation/create.go:174]: Ah ok, didn't realise destination will show as well on the help. sgtm in that case, thanks!

File: internal/cli/atlas/livemigrations/create.go [hunk 7]
```diff
@@ -15,34 +15,17 @@
...
-	migrationHosts              []string
-	sourceCACertificatePath     string
-	sourceClusterName           string
-	sourceProjectID             string
-	sourceSSL                   bool
-	sourceManagedAuthentication bool
-	sourceUsername              string
-	sourcePassword              string
-	store                       store.LiveMigrationCreator
+	options.LiveMigrationsOpts
```
Comment (colm-quinn) [inline internal/cli/atlas/livemigrations/create.go:27]: 💯

File: internal/cli/atlas/livemigrations/options/live_migrations_opts.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	if !opts.SourceManagedAuthentication && opts.SourceUsername == "" {
+		return fmt.Errorf("MongoDB Automation is not managing authentication, --%s must be set", flag.LiveMigrationSourceUsername)
+	}
+	if opts.SourceCACertificatePath != "" {
+		opts.SourceSSL = true
+	}
+	return nil
+}
+
+func (opts *LiveMigrationsOpts) GenerateFlags(cmd *cobra.Command) {
```
Comment (colm-quinn) [inline internal/cli/atlas/livemigrations/options/live_migrations_opts.go:129]: Very nice

**Review by colm-quinn — APPROVED** [review]
Great changes here. LGTM.

File: internal/cli/atlas/livemigrations/options/live_migrations_opts.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+	SourceCACertificatePath     string
+	SourceClusterName           string
+	SourceProjectID             string
+	SourceSSL                   bool
+	SourceManagedAuthentication bool
+	SourceUsername              string
+	SourcePassword              string
+}
+
+func (opts *LiveMigrationsOpts) CreateRequest() *mongodbatlas.LiveMigration {
```
Comment (gssbzn) [inline internal/cli/atlas/livemigrations/options/live_migrations_opts.go:44]: [not-blocking] new?
```suggestion
func (opts *LiveMigrationsOpts) NewCreateRequest() *mongodbatlas.LiveMigration {
```

File: internal/cli/atlas/livemigrations/options/live_migrations_opts.go (hunk 8 above)
Comment (gssbzn) [reply internal/cli/atlas/livemigrations/options/live_migrations_opts.go:129]: 💯

**Review by gssbzn — APPROVED** [review]
LGTM

**Review by gssbzn — APPROVED** [review]
LGTM thanks for all the work here, those changes for the shared struct look amazing

**Review by gssbzn — APPROVED** [review]

**Review by melissamahoney-mongodb — APPROVED** [review]

**Pull request merged**